### DISCUSSION TITLE
Comments on the robotSupervisorSchemeTutorial

### DISCUSSION THREAD

**Discussion by tsampazk** [aidudezzz/deepbots-tutorials#12] [General]
[Link to the tutorial.](https://github.com/aidudezzz/deepbots-tutorials/tree/master/robotSupervisorSchemeTutorial)

**Comment by bicrxm** [comment]
Hi this is Bikram. I want to contribute in the project for gsoc-21. Kindly let me know on how should I proceed.
Resume here:
[https://drive.google.com/file/d/1A2jD_j1gMdtuGTLgIUvtja7xt7777ADF/view](url)

**Reply by eakirtas** [reply]
Hello @bickrombishsass, first of all you should introduce your self on GFOSS mailing list. I would recommend to reproduce and run the examples that can be found in deepworlds repository. After all, you can start looking on deepbots issues or enhancements for the examples. There are a lot of space for improvements. You can ping us in order to help and guide you both on github and GFOSS matrix channel.

**Reply by eakirtas** [reply]
For instance, a good enhancement for the deepworlds examples will be to apply a [stable-baseline](https://stable-baselines.readthedocs.io/en/master/) algorithm in any example

**Comment by ringpy** [comment]
Hi, we are trying to reproduce the find and avoid in our world. We are not able to transmit the created message from epuck to supervisor. Do you have a quick document on this program and how to use the json scripts for emitter/receiver.

**Comment by Mouses94** [comment]
Hi, I am writing a paper with the goal of implementing the stable_baselines algorithms in a Webots environment (using the IRB4600). The first step for me is now this tutorial.
I have gone through the tutorial but now I always get the following error message:

INFO: robotSupervisorController: Starting controller: python.exe -u robotSupervisorController.py
Traceback (most recent call last):
  File "D:\Webots Projekte\controllers\robotSupervisorController\robotSupervisorController.py", line 88, in <module>
    env = CartpoleRobot()
  File "D:\Webots Projekte\controllers\robotSupervisorController\robotSupervisorController.py", line 16, in __init__
    self.robot = self.getSelf()  # Grab the robot reference from the supervisor to access various robot methods
AttributeError: 'CartpoleRobot' object has no attribute 'getSelf'
WARNING: 'robotSupervisorController' Controller beendet mit Status: 1

Another question: 
I already have the framework for the IRB4600, the motors and armChain, and above that the position sensors, I want to implement with the package ikpy as shown in the example Inverse kinematic. 
For this I need the supervisor method: supervisor.getUrdf().encode('utf-8')

However, I cannot implement this at the same time as the RobotSupervisor. Is there another solution for this?

Is there already an example project for this?

Many thanks

**Reply by eakirtas** [reply]
Hi @Mouses94, which deepbots version are you using?

**Reply by Mouses94** [reply]
Hi @ManosMagnus, thank you for the fast reply!
I used deepbots 0.1.2, now I've unintsalled it and installed the recommended 0.1.3 Version.
Now it works! 
Thank you!

**Reply by Mouses94** [reply]
However, now I have another Issue with the Deepbots Framework. 

I have written a Controller for the IRB4600 and wanted to run it with the Deepbots Framework. 

But theres now this Error:
```
Traceback (most recent call last):
  File "D:\Webots Projekte\controllers\ReachIRB\ReachIRB.py", line 293, in <module>
    selectedAction, actionProb = agent.work(observation, type_="selectAction")
  File "D:\Webots Projekte\controllers\ReachIRB\PPO_agent.py", line 72, in work
    c = Categorical(action_prob)
  File "C:\Users\Office\AppData\Local\Programs\Python\Python39\lib\site-packages\torch\distributions\categorical.py", line 64, in __init__
    super(Categorical, self).__init__(batch_shape, validate_args=validate_args)
  File "C:\Users\Office\AppData\Local\Programs\Python\Python39\lib\site-packages\torch\distributions\distribution.py", line 53, in __init__
    raise ValueError("The parameter {} has invalid values".format(param))
ValueError: The parameter probs has invalid values
```

What can I do about it? 
Does it have anything to do with the observation space, or where could the Error come from?

Thank you

**Reply by eakirtas** [reply]
What type are the observation, could you please add some samples? Pytorch seems to waits `Categorical` values does this makes sense?

**Reply by caozx1110** [reply]
> 

seems like only deepbots==0.1.3.dev2 can work

**Reply by tsampazk** [reply]
> > 
> 
> seems like only deepbots==0.1.3.dev2 can work

Hello, @caozx1110. The discussion you answered to is quite old now (May 2021). 

Right now, indeed the tutorial is intended for version 0.1.3.dev2 as stated [here.](https://github.com/aidudezzz/deepbots-tutorials/tree/master/robotSupervisorSchemeTutorial#prerequisites) I am currently updating the tutorial for the latest changes in Webots, fixing some issues with rotations, etc.

**Reply by caozx1110** [reply]
> 

ok, thanks for your response

**Comment by Kai-Mansfield** [comment]
Hi, great tutorial! Are there definitions for some of the methods, like ```getSelf()```, ```getDevice()```, ```getFromDef()```, etc. anywhere please? I tried searching the Deepbots and Webots github pages, but couldn't find the definitions or their methods and attributes. I found a reference to them in [webots\docs\reference\supervisor.md](https://github.com/cyberbotics/webots/blob/8e427f2cc4209ac854edda142e673896ccbc83ff/docs/reference/supervisor.md), but their methods and attributes aren't included. Thank you for your time :)

**Reply by tsampazk** [reply]
Hey @Kai-Mansfield, thanks! 

As far as i know, you can work off of the documentation [here](https://cyberbotics.com/doc/reference/supervisor#wb_supervisor_node_get_self) which is the same as the one you linked but in a more readable format. In most cases, the Webots python bindings are more or less self-explanatory, and if needed, most information exists on the documentation texts describing the methods.This and a little bit of experimenting usually get me the results i need.

If you still face problems you can visit the [official Webots discord (invite link)](https://discordapp.com/invite/nTWbN9m) and search there for similar questions or even ask a question yourself. Usually, one of Webot's developers will answer.

As a side note there is a [new python API in the works for Webots](https://github.com/cyberbotics/webots/pull/3801), which will probably make it a lot easier to work with python.

**Reply by Kai-Mansfield** [reply]
Hi @tsampazk, thank you for such a quick response. Excellent, I will have a look at the links, thank you!

**Comment by Kaijun101** [comment]
Hi, thanks for the awesome tutorial! However, when I run the tutorial I got the error, ModuleNotFoundError: No module named 'controller'. Would kindly require your assistance on this. The error is from supervisor_env.py.

**Reply by tsampazk** [reply]
Hello @Kaijun101 ! Could you please tell me what version of deepbots you have installed?

**Reply by Kaijun101** [reply]
Hi @tsampazk, thanks for your quick reply. I'm currently on 0.1.3 dev2.

**Reply by Kaijun101** [reply]
Hi @tsampazk , thanks for your help, I managed to figure out that there was a reference to a older version of deepbots using conda.

**Reply by tsampazk** [reply]
Sorry @Kaijun101 i didnt get a chance to look into it, good thing you managed to resolve the issue yourself! Thanks for letting me know. 😄

**Reply by Kaijun101** [reply]
@tsampazk how can I implement a RL algo with a let say the built in salamander bot in webots? I dont quite follow how to implement this. Would appreciate your guidance on this hahaha?

**Reply by tsampazk** [reply]
The process should be similar to the process followed on the tutorials on this repo. You already have the robot and the world from the example built in Webots, so you need to create an RL environment and figure out what task you want to solve. E.g. you might want the robot to navigate to a target position. This would be up to you to set it up and then use deepbots to create an environment class and implement the methods specifically for your environment, i.e. the reward function, etc., as seen in the tutorials for cartpole. You also need to provide the actual RL agent implementation of your choice and train it with the environment you set up. 

Deepbots as a middleware framework serves to help you implement the gym-style environment and make it compatible with already implemented agents, for the specific task and world you already have in mind.

**Reply by Kaijun101** [reply]
Hi @tsampazk, thanks for your guidance, thats great! For the robot, how can I get the global position of the 2 segments with the leg attached to the robot (refer to attached image)?
![salamander_robot](https://user-images.githubusercontent.com/90491438/195555770-6db96bfa-05ee-4ebc-965c-d91b3cb84fe3.png)

**Reply by tsampazk** [reply]
No problem @Kaijun101, happy to help! :smiley:  You have to do something similar to [this](https://github.com/aidudezzz/deepworlds/blob/17fcf949f14349136b0574a6d7e31253f4359a13/examples/find_and_avoid/controllers/supervisor/utilities.py#L16), granted you have references to the two nodes. Please refer to [webots documentation](https://cyberbotics.com/doc/reference/transform) for more information, as this is out of the scope of this tutorial.

**Reply by Kaijun101** [reply]
Thanks a lot @tsampazk! It appears that there is no references for the 2 nodes, according to [this](https://github.com/cyberbotics/webots/blob/master/projects/robots/epfl/biorob/protos/Salamander.proto). Am I right?

**Reply by tsampazk** [reply]
There are actually two nodes that look like the ones you are interested in, SEGMENT_THORAX and SEGMENT_PELVIS. You can use [get from def](https://cyberbotics.com/doc/reference/supervisor?tab-language=python#wb_supervisor_node_get_from_def) from a supervisor node with these two names and see if you actually get the correct references. I am not 100% on how this will work as the robot is defined as a PROTO.

I think you have a better chance to solve your Webots-related questions in the [official discord.](https://discord.com/invite/nTWbN9m)

**Reply by Kaijun101** [reply]
Then another question is I define my action space to consists of 11 actions, but the action output is a bool instead of an array of 11 values.
self.action_space = Box(low=np.array([-2.0, -2.0, -2.0, -1.5, -1, -1, -1, -1, -1, -1, -1]), high=np.array([2.0, 2.0, 2.0, 1.5, 1, 1, 1, 1, 1, 1, 1]), dtype=np.float64)

**Reply by tsampazk** [reply]
What do you mean the action output is a bool? Whose action output is that? Your RL agent's? You would need to either define an action space that is compatible with your agent and modify the environment accordingly, or use an agent that has action output compatible with your environment, depending on the use case.

**Reply by Kaijun101** [reply]
So I define my action space as shown above to 11 action outputs. But when at the def apply_action(self, action) portion, the action into the function is a bool instead of an array of 11 action values. I assume that action is from the network, right?

**Reply by tsampazk** [reply]
Well that would be my guess!

**Reply by Kaijun101** [reply]
I was using the PPO algo from the tutorial, any ideas how output multiple actions @tsampazk hahaha?

**Reply by tsampazk** [reply]
If your problem/environment uses a continuous action space, then the PPO agent used in the tutorial is not suitable as it is implemented for discrete action space. 

There's a wealth of continuous action space agent implementations online, as well as libraries like stable-baselines that you can use.

**Reply by Kaijun101** [reply]
Alright, and really thanks a lot @tsampazk  for all your support! Appreciate it definitely.

**Reply by tsampazk** [reply]
No problem @Kaijun101, happy to help!

**Comment by ChrisSim01** [comment]
Great, crystal clear tutorial!  Many thanks indeed for your work.
I plan to use this as the basis for my future work.  I have been using Webots for a while and this tutorial is just what I needed as a guide into RL.
If you can offer any advice into the following it would be greatly appreciated:
-saving the ANN state so that it can be loaded again from file, without the need to retrain.
-restart the training from an already trained n/w, perhaps on a slightly different task?
Best regards,  Chris

**Reply by tsampazk** [reply]
Hello @ChrisSim01, thank your for your kind words, very happy to hear that the tutorial proved useful for you! 😄 

As for your first question, you can find the agent save/load methods [here](https://github.com/aidudezzz/deepbots-tutorials/blob/48f76ecc9791d6b9cfc415623b4ad30d208efc9c/robotSupervisorSchemeTutorial/full_project/controllers/robot_supervisor_controller/PPO_agent.py#L93-L113), which are basically saving the actor/critic neural nets. They can be called whenever you wish, per some number of episodes, every episode or at the end of training to save the agent.

As long as you saved the models once, you can load them after the agent is initialized [here](https://github.com/aidudezzz/deepbots-tutorials/blob/48f76ecc9791d6b9cfc415623b4ad30d208efc9c/robotSupervisorSchemeTutorial/full_project/controllers/robot_supervisor_controller/robot_supervisor_controller.py#L95). To continue training on a different task with the same agent, you can modify your environment, webots worlds, etc. and just load the agent as discussed.

Hope this helps, let me know if you have other questions!

**Comment by ChrisSim01** [comment]
Awesome, thank you so much!

**Comment by ChrisSim01** [comment]
Hi again,

You could make an online course out of this tutorial via Udemy.com and make
money from it!
Just an idea,

Best regards,
Chris

**Reply by tsampazk** [reply]
Good idea Chris, i will definitely consider it! Thank you.

**Comment by ChrisSim01** [comment]
Hi again Kostas,

I do have one further question please:  is there any way to 'lock', or
encapsulate, a neural network module once it is learned, so that it is used
by, but not changed by further neural network training?   This would allow
tasks to be broken down into subtasks, and chaining of the subtasks
together I think.

I'm thinking in the style of
https://nn.cs.utexas.edu/downloads/papers/lessin.gecco13.pdf

Best regards,

Chris

**Reply by tsampazk** [reply]
Very interesting paper! 

I think that what you are describing can be achieved by "freezing" specific layers of your networks, something along the lines of [this](https://discuss.pytorch.org/t/how-the-pytorch-freeze-network-in-some-layers-only-the-rest-of-the-training/7088).

**Comment by ChrisSim01** [comment]
Thanks again Kostas.  I'm on a steep learning curve!

**Reply by tsampazk** [reply]
You sure are, but with persistence everything is possible!
